python/GafferImageUI/ImageViewUI.py

```diff
@@ -1300,49 +1350,217 @@ def __toggle( self, *args ):
+
+			# TODO - why does the pinning dropdown use collapsibleArrowDown instead of menuIndicator?
+			self.__menuButton = GafferUI.Button( image="menuIndicator.png", hasFrame=False, highlightOnOver=False )
+			self.__menuButton._qtWidget().setObjectName( "menuDownArrow" )
+			self.__menuButton.buttonPressSignal().connect( Gaffer.WeakMethod( self.__showEditorFocusMenu ), scoped=False )
+
+		self.addChild( row )
+
+		self.buttonPressSignal().connect( Gaffer.WeakMethod( self.__showEditorFocusMenu ), scoped=False )
+
+		self._setState( self.__defaultNodeSet, 1 )
+
+	def _setState( self, nodeSet, outputIndex ):
+		self.__nodeSet = nodeSet
+		self.__catalogueOutput = outputIndex
+		self.__memberAddedConnection = self.__nodeSet.memberAddedSignal().connect(
+			Gaffer.WeakMethod( self.update ), scoped = True
 		)
-		self.__button._qtWidget().setMaximumWidth( 25 )
+		self.__memberRemovedConnection = self.__nodeSet.memberRemovedSignal().connect(
+			Gaffer.WeakMethod( self.update ), scoped = True
+		)
+		self.update()
 
-		GafferUI.PlugValueWidget.__init__( self, self.__button, plug, **kw )
+	# TODO - hotkeys?
+	"""@staticmethod
+	def editorKeyPress( editor, event ) :
 
-		self._updateFromPlug()
+		if not isinstance( editor, GafferUI.NodeSetEditor ) :
+			return False
 
-	def _updateFromPlug( self ) :
+		if event.key == "N" and not event.modifiers :
+			_PinningWidget.__followNodeSelection( editor )
+			return True
+		elif event.key == "QuoteLeft" :
+			_PinningWidget.__followFocusNode( editor )
+			return True
+		elif event.key == "P" and not event.modifiers :
+			_PinningWidget.__pinToNodeSelection( editor )
+			return True
 
-		with Gaffer.Context() :
-			v = self.getPlug().getValue()
+		return False
+	"""
+
+	def update( self, *unused ) :
+		compareImage = None
+		
+		if self.__nodeSet is self.__defaultNodeSet:
+			compareImage = self.__node["in"]
+		elif len( self.__nodeSet ):
+			for plug in Gaffer.Plug.RecursiveOutputRange( self.__nodeSet[-1] ) :
+				if not plug.getName().startswith( "__" ) and type( plug ) == GafferImage.ImagePlug:
+					compareImage = plug
+					break
+
+		self.__node["compare"]["image"].setInput( compareImage )
+		self.__node["compare"]["catalogueOutput"].setValue( "output:%i" % self.__catalogueOutput if self.__catalogueOutput > 0 else "" )
+
+		# Icon
+
+		if self.__catalogueOutput > 0:
+			icon = "catalogueOutput%i.png" % self.__catalogueOutput
+		elif self.__nodeSet.isSame( self.__scriptNode.selection() ) :
+			icon = "nodeSetNodeSelection.png"
+		elif self.__nodeSet.isSame( self.__scriptNode.focusSet() ) :
+			icon = "nodeSetFocusNode.png"
+		else :
+			icon = "nodeSet%s.png"  % self.__nodeSet.__class__.__name__
 
-		icon = "catalogueOutput%s.png"%v[-1]
-		self.__button.setImage( icon )
+		# Make sure we don't break if an image is missing when setImage throws
+		try :
+			self.__icon.setImage( icon )
+		except Exception as e :
+			sys.stderr.write( str(e) )
```

Comment:
It feels like it would be hard to guarantee that no users of Image are expecting to get an exception when it fails to construct ( and what state is that supposed to leave it in? ).  A reasonable compromise seems to be moving the code from CompoundEditor ( where I copied it from ) to Button ( as well as switching to msg() ).

Reply:
> It feels like it would be hard to guarantee that no users of Image are expecting to get an exception when it fails to construct

It feels like a pretty reasonable assumption to me. What use cases are you envisaging where the exception handling would be any more than printing a message and carrying on?

> and what state is that supposed to leave it in?

Showing the warning icon maybe? Or just nothing? The reason I was thinking in this direction was that we recently had a case where Gaffer was completely crippled because a custom tool was missing its icon, and an unhandled exception prevented the UI from loading...

> A reasonable compromise seems to be moving the code from CompoundEditor

That's better, and I'll merge it, but I still think there's a good case for doing this even more centrally.